druid/src/tests/harness.rs

```diff
@@ -57,36 +61,104 @@ struct Inner<T> {
+    /// render_context to the `render_context_closure` callback.
+    ///
+    /// For lifetime reasons™, we cannot just make a harness. It's complicated.
+    /// I tried my best.
+    ///
+    /// The with_render version of `create` also has a callback that can be used
+    /// to save or inspect the painted widget
+    ///
+    /// # Usage
+    ///
+    /// The create functions are used to test a widget, it takes a `root` widget and a data structure
+    /// and creates a `Harness` that is interacted with via the `harness_closure` callback and the final render of
+    /// the widget can be inspected with the `render_context_closure` callback.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - A structure that matches the type of the widget and that will be
+    ///  passed to `harness_closure` callbacks `Harness` structure.
+    ///
+    /// * `root` - The widget under test
+    ///
+    /// * `shape` - The shape of the render_context in the `Harness` structure
+    ///
+    /// * `harness_closure` - A closure used to interact with the widget undertest through the
+    /// `Harness` structure.
+    ///
+    /// * `render_context_closure` - A closure used to inspect the final render_context via the `TargetGard` structure.
+    ///
+    pub fn create_with_render(
+        data: T,
+        root: impl Widget<T> + 'static,
+        window_size: Size,
+        mut harness_closure: impl FnMut(&mut Harness<T>),
+        mut render_context_closure: impl FnMut(TargetGuard),
+    ) {
         let mut device = Device::new().expect("harness failed to get device");
-        let target = device.bitmap_target(400, 400, 2.).expect("bitmap_target");
+        let target = device
+            .bitmap_target(window_size.width as usize, window_size.height as usize, 1.0)
+            .expect("bitmap_target");
         let mut target = TargetGuard(Some(target));
-        let piet = target.0.as_mut().unwrap().render_context();
-
-        let desc = WindowDesc::new(|| root);
-        let window = Window::new(WindowId::next(), Default::default(), desc);
-
-        let inner = Inner {
-            data,
-            env: theme::init(),
-            window,
-            cmds: Default::default(),
-        };
-
-        let mut harness = Harness {
-            piet,
-            inner,
-            window_size: DEFAULT_SIZE,
-        };
-        f(&mut harness);
+        {
+            let piet = target.0.as_mut().unwrap().render_context();
+
+            let desc = WindowDesc::new(|| root);
+            let window = Window::new(WindowId::next(), Default::default(), desc);
+
+            let inner = Inner {
+                data,
+                env: theme::init(),
+                window,
+                cmds: Default::default(),
+            };
+
+            let mut harness = Harness {
+                piet,
+                inner,
+                window_size,
+            };
+            harness_closure(&mut harness);
```

Comment:
The solution to the crash should be to add the following line after calling `harness_closure`.

```rust
harness.piet.finish().expect("piet finish failed");
```

Reply:
Thank you, I will go try this out soon